### PR TITLE
NF: `create_sibling_github` (see gh-335)

### PR DESCRIPTION
**Pull request by mih**
This is a working preview. Some aspects will be later test(ed/able). But
I do not plan to test the actual interaction with the Github API.

Many possible features are missing (customizable repo name templating,
...). It is unlikely that this PR will get them, because my focus is on
addressing the core of #335

TODO for #335:

- [x] enhance `add_sibling` to be able to configure a special-remote (type git)
- [x] enhance `add_sibling` to configure a publish dependency on another sibling
- [x] make `publish` to honor a publish dependency a TODO in #1037 

When this is done, one can create two siblings (one regular, one on github), and afterwards simply `datalad publish github` to have all data and handles moved to a server and afterward all handles to github with a autoenabled special remote, so anyone installing from github gets data too.

### CODE REVIEW

**Comment by yarikoptic** [comment]
tried on a plain datasets.datalad.org  clone with few datasets fetched... it seems to try to enable github remote for annex -- but it shouldn't since it should actually be annex=ignore
```shell
(git)hopa:/tmp/datasets.datalad.org[master]git
$> datalad create-sibling-github --github-organization datalad-collection-1 -r --existing=reconfigure datasets.datalad.org
...
i> 
2016-10-19 16:16:02,825 [ERROR  ] Failed to run ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'annex', 'enableremote', 'github'] under u'/tmp/datasets.datalad.org/crcns/aa-2'. Exit code=1. out=enableremote github failed
|  err=git-annex: enableremote: 1 failed
|  
2016-10-19 16:16:02,825 [INFO   ] Failed to enable annex remote github, could be a pure git 
2016-10-19 16:16:02,877 [ERROR  ] Failed to run ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'annex', 'enableremote', 'github'] under u'/tmp/datasets.datalad.org/kaggle/melbourne-university-seizure-prediction'. Exit code=1. out=enableremote github failed
|  err=git-annex: enableremote: 1 failed
|  
2016-10-19 16:16:02,877 [INFO   ] Failed to enable annex remote github, could be a pure git 
'git://github.com/datalad-collection-1/datasets.datalad.org.git' configured as sibling 'github' for <Dataset path=/tmp/datasets.datalad.org>
'git://github.com/datalad-collection-1/datasets.datalad.org-crcns.git' configured as sibling 'github' for <Dataset path=/tmp/datasets.datalad.org/crcns>
'git://github.com/datalad-collection-1/datasets.datalad.org-crcns-aa-2.git' configured as sibling 'github' for <Dataset path=/tmp/datasets.datalad.org/crcns/aa-2>
'git://github.com/datalad-collection-1/datasets.datalad.org-kaggle.git' configured as sibling 'github' for <Dataset path=/tmp/datasets.datalad.org/kaggle>
'git://github.com/datalad-collection-1/datasets.datalad.org-kaggle-melbourne-university-seizure-prediction.git' configured as sibling 'github' for <Dataset path=/tmp/datasets.datalad.org/kaggle/melbourne-university-seizure-prediction>
```

**Comment by yarikoptic** [comment]
how difficult would be to implement 'replace' strategy for existing?

**Comment by yarikoptic** [comment]
gy -- with such a huge patch (on top of #1011) 
```
--- a/datalad/distribution/utils.py
+++ b/datalad/distribution/utils.py
@@ -269,7 +269,12 @@ def _get_flexible_source_candidates(src, base_url=None):
         else:
             base_path = ri.path
             base_suffix = ''
-        ri.path = normpath(opj(base_path, src, base_suffix))
+        if ri.hostname == 'github.com':
+            # It must be a flat hierarchy
+            assert(base_suffix == '')
+            ri.path = "%s-%s" % (base_path, src)
+        else:
+            ri.path = normpath(opj(base_path, src, base_suffix))
 
     src = str(ri)
 
```
we should be able to install/get those published from github (see e.g. a limited collection of those which I have installed prior adding a sibling at  https://github.com/datalad-collection-1)

**Comment by yarikoptic** [comment]
I would still prefer if double slash `--` was used to separate levels in a flat hierarchy (as we hypothesized in the proposal).  This is more likely to avoid any collision (e.g. between `n1/n2-suf`, `n1/n2/suf`, and `n1-n2/suf`).  Moreover we could place that check/restriction during `create` to not allow `--` (or make it interactive to get a confirmation to allow) in the names.

**Comment by mih** [comment]
Thanks for the fixes!

> it seems to try to enable github remote for annex -- but it shouldn't since it should actually be annex=ignore

Will try to figure out where to pass this -- in any case annex sees it eventually and sets this by itself.

> how difficult would be to implement 'replace' strategy for existing?

Depends on what "replace" should do? If it is to wipe out an existing repo, it requires user/passwd auth (token is not enough). Otherwise it has full access and can destroy anything. Not sure how many warning signs I would want to go with that -- it is not only a repo that is gone in this case (which could be brought back easily in many cases), but also issues, wikis, access configurations, ....).

> I would still prefer if double slash -- was used to separate levels in a flat hierarchy (as we hypothesized in the proposal). This is more likely to avoid any collision (e.g. between n1/n2-suf, n1/n2/suf, and n1-n2/suf). Moreover we could place that check/restriction during create to not allow -- (or make it interactive to get a confirmation to allow) in the names.

This is all doable when any templating is introduced. I for one will *not* use any double dashes, and not use the current simplistic implementation either. Each repo is (and should be) standalone on github, a user does not need to care of any *possibly* existing hierarchy, any particular repo can be part of any number of hierarchies at the same time, and github does a good job of linking interconnected subdatasets to each other already.

**Comment by mih** [comment]
Related: #1035

**Comment by mih** [comment]
Seems to work. Will merge this now and @bpoldrack will finish the job as part of #1037 (thanks!).

**Comment by yarikoptic** [comment]
> I for one will not use any double dashes, and not use the current simplistic implementation either

Hm... Then for whom this implementation is if neither of us will/could use it? :-)
Or on another side - if you don't use anyways it lets use -- and then adjust through templating :-)

If for sfn, if not crcns we could demo having the same repos on datasets.d.o and github, and have them usable/installable, it would be great

**Comment by yarikoptic** [comment]
As for replace - could force push existing and then force push removal of old unique branches?

File: datalad/distribution/tests/test_create_github.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from nose.tools import assert_raises, assert_in, assert_true, assert_false, \
+    assert_not_in, assert_equal
+from nose import SkipTest
+
+
+try:
+    import github as gh
+except ImportError:
+    # make sure that the command complains too
+    assert_raises(MissingExternalDependency, create_sibling_github, 'some')
```
Comment (yarikoptic) [inline datalad/distribution/tests/test_create_github.py]: Lovely! Although I would not exercise assert right at import time -  make it into a new test_ function right here and nose will pick it up I i guess.

File: datalad/distribution/create_sibling_github.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    @staticmethod
+    def result_renderer_cmdline(res, args):
+        from datalad.ui import ui
+        res = assure_list(res)
+        if args.dryrun:
+            ui.message('DRYRUN -- Anticipated results:')
+        if not len(res):
+            ui.message("Nothing done")
+        else:
+            for d, url, existed in res:
```
Comment (yarikoptic) [inline datalad/distribution/create_sibling_github.py]: I think output is nowhere described... Btw shouldn't it be an ordered (if order matters) dict with d as a key?  I guess I would be to look also into regular create-sibling..  beasts should be similar

File: datalad/distribution/create_sibling_github.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+        reponame=Parameter(
+            args=('reponame',),
+            metavar='REPONAME',
+            doc="""Github repository name. When operating recursively,
+            a suffix will be appended to this name for each subdataset""",
+            constraints=EnsureStr()),
+        recursive=recursion_flag,
+        recursion_limit=recursion_limit,
+        sibling_name=Parameter(
+            args=('--sibling-name',),
```
Comment (mih) [inline datalad/distribution/create_sibling_github.py]: I don't agree. It makes no sense to do it this way for this command.

File: datalad/distribution/create_sibling_github.py (hunk 2 above)
Comment (mih) [inline datalad/distribution/create_sibling_github.py]: Fine, but the nature of this one is different. The old one doesn't even have --dryrun. one thing at a time.

**Comment by mih** [comment]
If you need to see double-dashes at once or nothing at all, go for it. This is just one piece in a bigger picture, and to me it is not yet the time to work on this because those hours are likely wasted. But I am not here to judge others ;-)

**Pull request merged**